wcfsetup/install/files/lib/acp/page/LicensePage.class.php

```diff
@@ -0,0 +1,637 @@
+<?php
+
+namespace wcf\acp\page;
+
+use CuyZ\Valinor\Mapper\Source\Source;
+use CuyZ\Valinor\MapperBuilder;
+use GuzzleHttp\Psr7\Request;
+use Laminas\Diactoros\Response\RedirectResponse;
+use wcf\acp\form\LicenseEditForm;
+use wcf\data\package\Package;
+use wcf\data\package\update\PackageUpdate;
+use wcf\data\package\update\PackageUpdateAction;
+use wcf\data\package\update\server\PackageUpdateServer;
+use wcf\page\AbstractPage;
+use wcf\system\database\util\PreparedStatementConditionBuilder;
+use wcf\system\io\HttpFactory;
+use wcf\system\request\LinkHandler;
+use wcf\system\WCF;
+
+/**
+ * Lists the licensed products and offers to install them.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.0
+ */
+final class LicensePage extends AbstractPage
+{
+    public $activeMenuItem = 'wcf.acp.menu.link.package.license';
+
+    public $neededPermissions = ['admin.configuration.package.canInstallPackage'];
+
+    private array $licenseData;
+
+    private int $licenseNumber;
+
+    private array $installedPackages;
+
+    private array $installablePackages = [];
+
+    private array $packageUpdates = [];
+
+    private PackageUpdateServer $updateServer;
+
+    public function readData()
+    {
+        parent::readData();
+
+        $this->updateServer = PackageUpdateServer::getWoltLabUpdateServer();
+
+        if (!$this->hasLicenseCredentials()) {
+            return new RedirectResponse(
+                LinkHandler::getInstance()->getControllerLink(LicenseEditForm::class),
+            );
+        }
+
+        (new PackageUpdateAction([], 'refreshDatabase'))->executeAction();
+
+        $this->licenseData = $this->fetchLicenseData();
+
+        $identifiers = \array_merge(
+            \array_keys($this->licenseData['woltlab']),
+            \array_keys($this->licenseData['pluginstore'])
+        );
+        $this->installedPackages = $this->getInstalledPackages($identifiers);
+
+        $identifiers = \array_diff($identifiers, \array_keys($this->installedPackages));
+        $identifiers = $this->removeUnknownPackages($identifiers);
+
+        if ($identifiers !== []) {
+            $this->installablePackages = $this->getInstallablePackages($identifiers);
+            $this->packageUpdates = $this->getPackageUpdates($this->installablePackages);
+        }
+
+        foreach (['woltlab', 'pluginstore'] as $type) {
+            $this->licenseData[$type] = \array_filter(
+                $this->licenseData[$type],
+                function (string $package) use ($identifiers) {
+                    if (isset($this->installedPackages[$package])) {
+                        return true;
+                    }
+
+                    return \in_array($package, $identifiers);
```

Comment:
Should this check for `$this->installablePackages`, instead of `$identifiers`? The code could probably benefit from *some* comments here, because the exact semantics are not immediately clear from the variable names and method names alone, because I'm not sure which packages / versions are pruned by `removeUnknownPackages` and which by `getInstallablePackages()`.

Reply:
The filter `removeUnknownPackages()` exists to remove packages that are returned by the API but are unknown to the package server. For example, the legacy styles offered during the Burning Board 3 era are still part of the licenses but these do not exist on any package server anymore.

I guess that `removeUnknownPackages()` is probably redundant at this point but it was required in an earlier iteration because there were some more steps between it and `getInstallablePackages()`.